### ISSUE TITLE
cups-browsed 1.24.0 does not create local queues for remote cups queues (cups-1.4.2)

### ISSUE THREAD

**Issue by zdohnal** [OpenPrinting/cups-filters#124]
Hi!

I tested the newest release - 1.24.0 - in our company environment, where print server is served by cups-1.4.2, and local queues are not created anymore. Here are the logs 
[cups-browsed-log.txt](https://github.com/OpenPrinting/cups-filters/files/3251248/cups-browsed-log.txt) .
From logs - it seems some of remote queues did not pass get-printer-attributes request at cups-browsed.c:6793 , thougn the last cups error is 'Success':

Jun 04 03:29:31 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:31 2019 Request for IPP attributes (get-printer-attributes) for printer with URI ipp://server_hostname:631/printers/brno4-tpbc-0th-north failed: Success
Jun 04 03:29:31 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:31 2019 get-printer-attributes IPP call failed on printer brno4-tpbc-0th-north (ipp://server_hostname:631/printers/brno4-tpbc-0th-north).

But what is interesting, some remote print queues passed that (or missed that request) and tried to load PPD, but with no luck:

Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 Failed reading file /var/cache/cups/cups-browsed-options-brno1-0th-cafe, probably no options recorded yet
Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 Print queue brno1-0th-cafe is for remote CUPS queue(s) and we get notifications from CUPS, using implicit class device URI implicitclass://brno1-0th-cafe/
Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 Unable to create PPD file: No such file or directory

IMO This two errors are connected - errno is not cleared after load_printer_options(), which ends with error, sets errno and it is again checked after ppdCreateFromIPP2() without clearing errno -> fails.
But the same print queue goes to record_printer_options() and produces following error:

Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 Unable to get PPD file for brno1-0th-cafe: The printer or class does not exist.
Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 lpstat: No destinations added.
Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 Unable to load PPD from local queue brno1-0th-cafe, queue seems to be raw.
Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 Queue has still jobs or CUPS error!
Jun 04 03:29:33 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:33 2019 ERROR: Failed disabling printer 'brno1-0th-cafe': The printer or class does not exist.

**Comment by tillkamppeter** [comment]
@deepak0405 can you have a look into this? Thanks.

**Comment by deepak0405** [comment]
I am trying to reproduce the errors. 
1. What is the Make Model of remote cups queue?
2. Can you please also send the output  and return status of 
```
ipptool -tv <URI> get-printer-attributes-2.0.test
  ```
https://github.com/steveathon/cups/blob/master/test/get-printer-attributes-2.0.test

**Comment by zdohnal** [comment]
@deepak0405 
1) Canon imageRunner C5185i Foomatic/Postscript , but it based on used PPD. In reality it is Canon imageRUNNER ADVANCE C3530i MFP 
2) ipptool -tv ipp://server_hostname:631/printers/brno4-tpbc-0th-north get-printer-attributes-2.0.test
[ipptool_output.txt](https://github.com/OpenPrinting/cups-filters/files/3252271/ipptool_output.txt)

The uri is uri of print queue on server, uri to device itself is different (socket://printer_ip:9100) - and client machine, server and printer are all in different subnets.

**Comment by tillkamppeter** [comment]
@zdohnal, the CUPS server server_hostname is not answering the standard get-printer-attributes IPP request, according to your ipptool_output.txt. As CUPS is supposed to fully conform to IPP 2.x this is not expected.
Most probably this is either a bug in CUPS or the server runs a stone-old CUPS version which only does IPP 1.x.
Could you tell which CUPS version your server server_hostname is running? Could you also attach its cupsd.conf and cups-files.conf to this bug report?
If you do not see any hint of a possible mis-configuration (not ver probable as both printing and the get-printer-attributes request go through IPP) please consider reporting a bug on CUPS upstream.

**Comment by zdohnal** [comment]
@tillkamppeter Ad cups version - please see issue subject, I explicitly defined version there - and yes, because it is old CUPS, it runs only IPP 1.x. 
Ad conf files - they are under jurisdiction of our IT team, so I do not have free access to them. But I asked them to send them to me.

**Comment by tillkamppeter** [comment]
@deepak0405, perhaps we should re-introduce the old clustering method for legacy server support. Not having this especially defeats the support for legacy CUPS browsing (as a client).
First, we need to find out with which version CUPS started to have IPP 2.x support (with supplying full capability info when getting a get-printer-attributes IPP request). If the first DNS-SD-enabled version already supports this (what I assume), we can continue treating all DNS-SD-discovered remote printers as we are treating them now and would treat printers discovered via CUPS (<= 1.5.x) legacy broadcast the old way, either with a raw client queue (no clustering possible) or with the download-server-PPD-and-modify-to-non-filtering-locally method (restricted clustering possible).
For integrating this legacy support in the current cups-browsed we need to take into account:
- Clustering of a legacy printer with downloaded server PPD with a modern network printer/CUPS queue with get-printer-attributes response would require a function which converts a PPD file into a get-printer-attributes response data structure.
- If creating a PPD->get-printer-attributes converter function is too complicated we must somehow manage to not be able to cluster modern with legacy. By default we should simply refuse to cluster modern with legacy, so a cluster started with a legacy printer should only accept further legacy printers and a cluster started with modern printers should only add further modern printers. 
- We could add one or more directives to cups-browsed.conf to select the behavior:
  - `ClusterModern`, `ClusterLegacy`: as `Cluster` but using the modern or legacy clustering method, independent of how the printers get discovered, `Cluster-modern` would reject legacy printers. This overrides the default behavior for clustering.
  - `ClusterMethod` sets the default clustering method. Default is `Auto` where the first printer determines which method to use, `Modern` uses the modern clustering and ignores/rejects legacy printers added to a cluster or a cluster formed from an already set up legacy printer, `Legacy` creates legacy clusters even if the first printers discovered are modern ones (so that legacy printers can get added later). Manual clusters with `ClusterModern` or `ClusterLegacy` override this default setting.

WDYT? How should we implement the legacy printer (IPP 1.x + CUPS browsing) support?

**Comment by tillkamppeter** [comment]
@zdohnal, as your servers have such an old CUPS (1.4.x), are you using cups-browsed's feature to create local queues based on legacy CUPS browsing (the feature Tim Waugh had added near the beginning of cups-browsed development in 2013)?
Do you only access single printers with each loacl queue or do you want to create clusters?

**Comment by zdohnal** [comment]
@tillkamppeter ad first question - AFAIK yes, cups-browsed is set to find remote shared printers by dnssd protocol and CUPS legacy protocol.
Ad second question - ideally both. I'm okay with accessing single printers with each local queue for my usual workflow, but IMO there will be requests for clustering feature too . 
e.g. our customers wanted clustering features in RHEL 7, so I am doing a rebase of cups-browsed to 1.13.4 without some other features, so there is quite high possibility they will rather want the feature than changing their infrastructure (buying IPP 2.x printers etc...)

**Comment by zdohnal** [comment]
Footnote:
There was a change in creating local queues from CUPS legacy broadcast since 1.14.0 - created queues started to get implicitclass backend, even when only one server (AFAIK) is broadcasting the queue - I thought it was simple change of behavior, but I'm not sure now - is it not a bug in the end?

**Comment by deepak0405** [comment]
@tillkamppeter
The DNS-SD support was introduced in cups1.2 and the IPP 2.x support was added in cups 1.4.

I think the method suggested by Till is the best way to solve the issue.

The new Clustering code does IPP requests three times:

1. While getting the attributes of the printer for generating the ppd file.

2. Getting the number of jobs on the server, while we are distributing the new job on the cluster.

3. Sending the job to the printer using IPP backend.

Will we be able to do the last two mentioned IPP request for the legacy printer ? 

Implementing a function to get printer attributes from ppd will be a difficult task. We will have to parse the ppd file and after parsing we may have to convert the ppd data into different form

for eg. 
The pagesize Letter in ppd file will be written something like 
 { 
   media-size = {x-dim = ...., y-dim = ....} , 
   media-type = ....,
 }

There is a function to convert the second format to the first one, but I am not aware of a function for the reverse work. We can do one of these two things here:

1. Write functions to convert the ppd options to IPP options.

2. If a legacy printer is in the cluster :
     a. Add function to ppdgenerator.c which reads the ppd file of the legacy printer and copies its options to the cluster ppd. In this way we don't have to convert the option.
     b. While selecting the printer, we need to write a different function to see whether the legacy printer supports the requested options. For modern printer we will match the job settings with the p->prattrs, and for legacy we will read the ppd file and see if the option is supported or not.

IMO, the second one will be easier.

**Comment by tillkamppeter** [comment]
@zdohnal, the implicitclass backend is needed for clustering and it is easier to handle when a single printer also uses implicitclass instead of changing when a second printer gets added or all but one printers of a cluster disappear. See also the `NEWS` file, changes for version 1.11.3.

**Comment by tillkamppeter** [comment]
@deepak0405, CUPS started with DNS-SD printer advertising only in 1.3.x but it does not change anything for us here as @zdohnal has CUPS 1.4.x on his servers. But having CUPS 1.4.x already doing IPP 2.x but not answering a get-printer-attributes IPP request properly looks like a bug in this old CUPS for me. @deepak0405 could you have a look into the IPP specs whether the IPP 2.0.x/2.1.x conformance is enough for the get-printer-attributes IPP response we expect?
I think polling number of jobs and sending off jobs via IPP should already work with old CUPS versions, these basic operations should already be available in IPP 1.x.
For the legacy printer integration in a cluster the function which turns a PPD into IPP attributes would be the most elegant and consistent solution, but we would need to use deprecated PPD functions of the CUPS library to not duplicate/re-invent code and as soon as CUPS removes them move them over into libcupsfilters. Can be a lot of code to maintain only to support legacy servers. But this method will also keep working when CUPS stops using PPDs for creating print queues and we then, too. So the local CUPS can be one which does not use PPDs any more and the server can use a stone-old CUPS which still does not answer get-printer-attributes requests properly.
Adding the options of a downloaded PPD from a legacy printer to the joint PPD of all modern printers in the cluster also has its quirks, as we need to find out whether the differently-named options/choices have the same meaning and also keep track of the original PPD to send the correct option settings when the legacy printer gets selected for a job. We also need to take into account when there is more than one legacy printer to add to the joint PPD or if there are only legacy printers and the joint PPD of the modern printers is therefore empty. Also this will stop working if CUPS removes PPD support as then we also do not create a PPD any more. With this method the selection of the destination printer can also get arbitrarily complicated as we have to convert the job's attributes into PPD options to see whether they match the legacy printers.
So for me it seems to be easier and more future proof to convert the downloaded PPDs into IPP attributes and then do all the rest as with a modern printer. Most straightforward.
The decision whether to proceed the standard way of a modern printer or to download the PPD we make based on the answer we get on a get-printer-attributes request. If it is usable we do the modern method, if not we try to download the PPD and only if this does not work we discard the printer as not supported.
WDYT?

**Comment by deepak0405** [comment]
@tillkamppeter 
The GET_PRINTER_ATTRIBUTES request is also supported by [IPP/1.1](https://tools.ietf.org/html/rfc2911). I checked in IPP/1.1 specification and the IPP request format was the same as we are doing. But the attributes lists in IPP/1.1 or  [IPP/2.x](https://ftp.pwg.org/pub/pwg/standards/std-ipp20-20151030-5100.12.pdf) doesn't have reference to media-col-database. One file related to [IPP](https://ftp.pwg.org/pub/pwg/candidates/cs-ippjobprinterext3v10-20120727-5100.13.pdf) mentions media-col-database. 
Whether the error can be because of requesting media-col-database? 

@zdohnal 
Can you please send the output of this [test](https://github.com/deepak0405/GSoC2018_documentation/blob/master/get-printer-attributes-2.0.test) file ?

**Comment by tillkamppeter** [comment]
@deepak0405, perhaps we need to work with a fallback. First, we do, as before, our get-printer-attributes request with "all" and "media-col-database", if we get a failure or broken data, we get into the fallback, which is the same get-printer-attributes request but only with "all" and no "media-col-database". This can even in addition to support for old printers/CUPS queues solve the Canon printer problem of issue #22 (or apple/cups#5512).

**Comment by zdohnal** [comment]
Hi @deepak0405 ! 
here are the logs 
[ipptool_output_without_mediacoldatabase.txt](https://github.com/OpenPrinting/cups-filters/files/3272271/ipptool_output_without_mediacoldatabase.txt)

**Comment by tillkamppeter** [comment]
Thanks, @zdohnal, this looks like that the fallback as I described in my previous comment would work with this output, as the output lists the available paper sizes ("media-supported") and input PDLs ("document-format-supported")  which is the minimum information for generating a valid PPD file, but it has also some point where it is not consistent, as under "job-settable-attributes-supported" the attributes "print-quality" and "printer-resolution" are listed and there are no "print-quality-supported" and "printer-resolution-supported" entries telling about valid choices. This is possibly a bug in the CUPS running on the server.

**Comment by zdohnal** [comment]
@tillkamppeter I checked clean CUPS 1.4.2 which Mike provided at that time and current upstream master - actually those two missing attrs were only in requested attributes field at that time and they were missing from load_ppd() function, where they are handled now in current upstream. So they are missing in the request because server did not take them from ppd file and was not meant to do it at that time...
Is their absence critical? Because IMO it is 'new feature' (in comparison of RHEL6 and cups-1.4.2) and we are cautious about not-breaking things in older RHEL, so I will need some good justification to get permission to fix it. (there is an initiative to upgrade print servers in the future, but I'm not sure about ETA...)

**Comment by tillkamppeter** [comment]
@zdohnal, the missing "print-quality-supported" and "printer-resolution-supported" (or generally missing "...-supported" if "..." is listed in "job-settable-attributes-supported") I consider a bug and not a feature addition as it breaks IPP standard compliance, so being legitimate to fix this in an already released distribution, but it is not a security bug or otherwise very severe. You could ask your release team whether they would accept the backport of the fix to this bug as it improves user experience a lot when modern clients (new OS versions, mobile devices) access printers on RHEL servers.
@deepak0405, indepenndent of above-mentioned bug i would recommend you to first implement as the very first approach a two-step poll for get-printer-attributes. First you do the "standard" way with attributes "all" and "media-col-database". If the result is unusable in any form you do the fallback of only using the "all" attribute. This should get you a usable but perhaps poor result. Implement this and let @zdohnal test. After having @zdohnal's answer we can still decide whether we need a PPD importer for legacy CUPS queue support or not.

**Comment by deepak0405** [comment]
@zdohnal 
I have generated a pull request #128. If an IPP request to obtain "media-col-database" and "all" fails, we do a fallback request, asking only "all" group name attributes.
Can you please test it ?

**Comment by zdohnal** [comment]
@deepak0405 Sure thing, I'll test it right away. Thank you for the patch!

**Comment by zdohnal** [comment]
@deepak0405 I cherry-picked your two commits, applied on 1.24.0 and tried it in Fedora rawhide. Unfortunately, there are no local queues. I'll send the logs on Monday.

**Comment by tillkamppeter** [comment]
@zdohnal, chank you very much. I have already found a bug in the patch which renders the fallback completely non-functional. Please see my comment on the pull request. You could try again with this bug corrected.

**Comment by zdohnal** [comment]
@tillkamppeter now I see - the fallback is very good idea, but it is unfortunate it is not set to 1 in any call of get_printer_attributes() :)

**Comment by tillkamppeter** [comment]
@zdohnal, in calls outside of get_printer_attributes() outside get_printer_attributes() itself the fallback mode is always set to 0 so that get_printer_attributes() does the standard ("all" + "media-col-database") first. If this fails, get_printer_attributes() calls itself with 1 as fallback mode, so that then the attempt with only "all" is initiated.
Did you test it? Did you apply the small correction of my comment in the pull request?

**Comment by zdohnal** [comment]
I'm sorry, I have not tested it yet - just arrived into the office. I'll post an update around the noon.

**Comment by zdohnal** [comment]
@tillkamppeter Actually I do not see any comment on pull request, so I looked into the code and IMO there is the other issue - when you call get_printer_attributes(uri, 1), you do not store the return value, so the output of get_printer_attributes is still NULL from previous call.
I tried to fix it, but messed up something else, so the queues are not created too.

**Comment by tillkamppeter** [comment]
@zdohnal, take deepak's repo on which he has created this pull request. There edit line 6659 in the file `utils/cups-browsed.c` by replacing
```
get_printer_attributes(uri,1);
```
with
```
response = get_printer_attributes(uri,1);
```
Now use the freshly built cups-browsed, either by terminating any other cups-browsed instance and then running `./cups-browsed` in the source tree of cups-filters or replacing the system's cups-browsed with `.libs/cups-browsed` from the source tree.
If you have made sure that you are running the new, fixed cups-browsed and still do not get client queues, switch cups-browsed into debug mode and provide us `/var/log/cups/cups-browsed_log`.

**Comment by deepak0405** [comment]
@zdohnal Please also upload the ppd file, if it is generated.

**Comment by zdohnal** [comment]
I'm sorry, tried pure deepak's version, but still no created queues - here is the log 
[cups-browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3300587/cups-browsed_log.txt) and there are no ppds in /etc/cups/ppd.

**Comment by tillkamppeter** [comment]
@zdohnal, could you do the get-ipp-attributes requests with only the "all" attribute using ipptool (copy file `/usr/share/cups/ipptool/get-printer-attributes-2.0.test`, remove "media-col-database" from the attributes and use this with ipptool). Do it preferably on all your remote printers, from the client on which you have the problem with cups-browsed. Attach the results here together with the printer's URIs. 

Especially I need the results for
ipp://server_name:631/printers/brno1-4th-cafe
ipp://server_name:631/printers/brno2-1st-hc-mazlik

**Comment by zdohnal** [comment]
Here are the logs of request 
[gpa-request-all.tar.gz](https://github.com/OpenPrinting/cups-filters/files/3301103/gpa-request-all.tar.gz) . I see in the logs:

 Get printer attributes using get-printer-attributes                  [FAIL]

even without media-col-database, is it connected?

**Comment by zdohnal** [comment]
Printer uris are in the logs (server_cname is canonical name of server, server_name is classic hostname).

**Comment by tillkamppeter** [comment]
@zdohnal, @deepak0405, the output seems to be a sufficient amount of data for building a PPD file, but ipptool is considering the operation failed, probably due to the following errors listed in the end of the file:
```
        "generated-natural-language-supported": Bad naturalLanguage value "en_US" - bad characters (RFC 2911 section 4.1.8).
        "natural-language-configured": Bad naturalLanguage value "en_US" - bad characters (RFC 2911 section 4.1.8).
```
These errors are probably not severe enough to prevent building a PPD, and they are most probably caused by a bug in this stone-old CUPS 1.4.2. A solution would be to let cups-browsed ignore certain errors of the get-printer-attributes request. In the output of ipptool we see that we get the data even with the errors, we only need to make cups-browsed accept it.
@deepak0405, could you try to change the error handling of the get-printer-attributes IPP request in cups-browsed, to
- ignore certain errors or even ignore all errors except a few which are known to make the output really unusable
- check the content of the response before considering the get-printer-attributes request failed
- when ignoring some or all errors try to biuild the PPD with the response and only if it fails do the fallback (no "media-col-databse" attribute) and only if this fails, too, drop the printer.

To check whether a response to a get-printer-attributes request is sufficient to build a PPD file run only this response (not the combination for the cluster) through the PPD generator and accept the response if this operation does not error. Discard the resulting PPD if the new printer is about to get added to a cluster.
@deepak0405, if you are in doubt whether a certain approach would work or not, you can get @zdohnal test several approaches, either by making the code in a way that @zdohnal only needs to do small changes at certain points of the code to get the alternatives being used or make several pull requests. And before submitting any new code to the current pull request or make it a base for other pull requests, please fix the bug I have mentioned in a separate commit. This makes testing and handling much easier.

**Comment by deepak0405** [comment]
@tillkamppeter There are a number of IPP status which can be returned. I am using this ref:
[IANA Repo](  https://www.iana.org/assignments/ipp-registrations/ipp-registrations.xhtml#ipp-registrations-11). I will read which errors can be ignored.

Also,
The log file shows that even though the request failed, our cupsLastErrorString() is Success. 
`Jun 04 03:29:25 host-client-ip cups-browsed[2420]: Tue Jun  4 03:29:25 2019 Request for IPP attributes (get-printer-attributes) for printer with URI ipp://server_hostname:631/printers/brno1-4th-eng-head failed: Success`
As already pointed by @zdohnal  the IPP Status is not set as cups Error status. I checked the code of IPP request in cups, and saw that the IPP return status is not set as cupsError anywhere. May be we can use ippErrorString(ipp_status) inplace of cupsLastErrorString() ?

**Comment by tillkamppeter** [comment]
I think using `ippErrorString(ipp_status)` inplace of `cupsLastErrorString()` is the correct thing here, it was probably overlooked when this code was initially created. Please do it.

**Comment by deepak0405** [comment]
Hi @tillkamppeter, @zdohnal I am working on the issue. It will take me few more days.

**Comment by PatriciaJH** [comment]
Hi all, I believe I may be seeing the same bug on a much simpler network -- no subnets, 2 printers, Cups running on my laptop.   Seeing it on Fedora 29 with all updates, and not on Linux Mint 19.1; manifests as lpq not seeing the default printer when I change it from the system Printers dialogue or from the Cups browser interface.   I am seeing repeated unresolved "cupsd[2925]: REQUEST localhost - - "POST / HTTP/1.1" 200 182 Renew-Subscription client-error-not-found" in my cups log. 

I can, however, see the printer using "lpstat -p -d" 
and set the default using (for instance) "lpoptions -d DCP-L2540DW"  
which also of course resolves the log errors -- but the default still can't be updated from the UI or browser.  NORMALLY I set this by the UI, and never need to touch the commandline. 

My printers are a Brother DCP-L2540DW  and an HP PhotoSmart 7150.  

Happy to follow and pull your code to test or help debug, or to file a new bug.

**Comment by tillkamppeter** [comment]
@PatriciaJH, for me this looks like a different issue. Could you start a new issue report and there especially tell:
- Which distribution version and which versions of CUPS and cups-filters are you using on the client? And which ones on the server?
- How did you set up your print queues on the server? Device URI? Are the printers set to be shared? Please attach the printer's PPD files (from `/etc/cups/ppd/`) to this bug report.
- Is the CUPS on the server set to share printers? Could you provide its `/etc/cups/cupsd.conf`?
- Are the client's queues auto-created by cups-browsed (if you stop cups-browsed they disappear, if you start cups-browsed they appear again)? Or are they auto-created by CUPS itself (only show under `lpstat -e`, not under `lpstat -v`)?
- Could you attach `/etc/cups/cupsd.conf` and `/etc/cups/cups-browsed.conf` of the client?
- Could you provide `/var/log/cups/error_log` of the CUPS of both client and server in debug mode? To switch CUPS to debug mode, run the command `cupsctl --debug-logging` on both the client and the server.
- Could you set cups-browsed of the client into debug mode (Put a line `DebugLogging file` into `/etc/cups/cups-browsed.conf` and restart cups-browsed) and then after some time, attach `/var/log/cups/cups-browsed_lo`?
Please do not package the files together, do not compress them, attach them one by one, rename them adding a `.txt` extension so that GitHub accepts them. Thanks.

**Comment by zdohnal** [comment]
@deepak0405 No problem, take your time - I'll let older cups-filters in Fedora for now. I will be on vacation next week, so if you need a testing, I will not reply until Monday 7th July.

**Comment by deepak0405** [comment]
@tillkamppeter 
The response without including "media-col-database" is NULL. I was trying to recreate the same environment so that I will be able to debug it better.

I used `run-stp-tests.sh`, luckily it was there in the cups 1.4.2 code. I was able to create print queues on port 8631 but those were not visible in cups-browsed running on the client.

a. The printers on port 631 were visible on the cups-browsed.
b. Command to create printer:

```
      lpadmin -p test3 -h localhost:8631 -m  "drv:///sample.drv/zebra.ppd" -o printer-is-shared=true -v 
       file:/home/deepak/Desktop/out
```
c. Output of `cupsctl -h localhost:8631`

```
_debug_logging=1
_remote_admin=0
_remote_any=0
_share_printers=1
_user_cancel_any=1
AccessLog=/tmp/cups-deepak/log/access_log
AccessLogLevel=actions
BrowseLocalProtocols=dnssd
CacheDir=/tmp/cups-deepak/share
DataDir=/tmp/cups-deepak/share
DocumentRoot=/home/deepak/GSoC2018_documentation/cups-release-1.4.2/doc
ErrorLog=/tmp/cups-deepak/log/error_log
FileDevice=yes
FontPath=/tmp/cups-deepak/share/fonts
LogTimeFormat=usecs
MaxLogSize=0
MaxSubscriptions=3
PageLog=/tmp/cups-deepak/log/page_log
PassEnv=LOCALEDIR
PreserveJobHistory=Yes
Printcap=
RequestRoot=/tmp/cups-deepak/spool
ServerBin=/tmp/cups-deepak/bin
ServerRoot=/tmp/cups-deepak
StateDir=/tmp/cups-deepak
TempDir=/tmp/cups-deepak/spool/temp
User=deepak
```
![cups queues](https://github.com/deepak0405/GSoC2018_documentation/blob/master/Screenshot%20from%202019-07-17%2009-12-52.png)

![shared queues](https://github.com/deepak0405/GSoC2018_documentation/blob/master/Screenshot%20from%202019-07-17%2009-13-11.png)

@tillkamppeter  How can I share these queues ?

**Comment by deepak0405** [comment]
@tillkamppeter
My earlier setup was : 
              Server ( VM ) : Ubuntu 19.04 and cups 1.4.2 running on port 8631
              Client: Ubuntu 19.04, latest cups-browsed and  cups 2.2.10.

I applied the patch on the avahi 0.7 code, installed and reloaded the systemctl service on the Server.
I checked the output of avahi-discover on client and avahi-browse -a on server, the printer on port 8631 was not broadcasted by the avahi service.

You wrote `if your run-stp-tests.sh test CUPS is on your local machine` , whether I should have run this script on client, currently I am running on server.

**Comment by tillkamppeter** [comment]
OK, what we need is a server which actually broadcasts its printers so that cups-browsed picks them up automatically. I do not know with which CUPS version the Avahi broadcasting support started. cups-browsed is also capable of picking up printers via the legacy CUPS broadcasts of old CUPS versions. Perhaps you need to ask @zdohnal what config he actually has.

**Comment by zdohnal** [comment]
@deepak0405 I asked our IT for access to the server (I expect you need server side cupsd.conf), but no answer yet... If you would like to know a specific setting in cupsd.conf on the server, maybe I can ask people who was in charge of company printers before current IT, but still I do not have access to whole cupsd.conf :( .

**Comment by logTom** [comment]
I also get `get-printer-attributes IPP call failed on printer konica_minolta_bizhub_c253_5e_92_c6_ (implicitclass://KONICA_MINOLTA_bizhub_C253_5E_92_C6_/).` executing cups-browsed on Solus with cups-filters 1.25.1.

**Comment by deepak0405** [comment]
@logTom 
* Which version of CUPS you are using ? Whether your printer is IPP Network Printer, if not what is the version of cups running on the server ?
* Can you please attach the error log of cups-browsed ?
* Please also send the output of 
```
       ipptool -tv <URI> get-printer-attributes-2.0.test
```
https://github.com/steveathon/cups/blob/master/test/get-printer-attributes-2.0.test

**Comment by logTom** [comment]
Cups is version 2.2.8. Should I try upgrading to the latest version 2.2.11+?
Or try cups-filters with the proposed fix https://github.com/OpenPrinting/cups-filters/pull/128?
The printer is a Konica Minolta C-253 network printer and yes it's kinda old.

The full error log is (edited /etc/cups/cups-browsed.conf and uncomment the line DebugLogging stderr):
```
sudo /usr/sbin/cups-browsed
Mon Aug 12 09:58:26 2019 Reading config: BrowseRemoteProtocols dnssd cups
Mon Aug 12 09:58:26 2019 Reading config: CreateIPPPrinterQueues All
Mon Aug 12 09:58:26 2019 No "Browse..." line at all, accept all servers ("BrowseOrder Deny,Allow").
Mon Aug 12 09:58:26 2019 main() in THREAD 139740392815168
Mon Aug 12 09:58:26 2019 cups-browsed: Creating http connection to local CUPS daemon: /run/cups/cups.sock:631
Mon Aug 12 09:58:26 2019 update_netifs() in THREAD 139740392815168
Mon Aug 12 09:58:26 2019 network interface lo: Local host name/address: 127.0.0.1
Mon Aug 12 09:58:26 2019 network interface lo: Local host name/address: localhost
Mon Aug 12 09:58:26 2019 network interface eno1: Local host name/address: 192.168.1.59
Mon Aug 12 09:58:26 2019 network interface eno1: Local host name/address: logmediapc002.local
Mon Aug 12 09:58:26 2019 Network interface eno1 at 192.168.1.59 for legacy CUPS browsing/broadcasting
Mon Aug 12 09:58:26 2019 network interface virbr0: Local host name/address: 192.168.122.1
Mon Aug 12 09:58:26 2019 Network interface virbr0 at 192.168.122.1 for legacy CUPS browsing/broadcasting
Mon Aug 12 09:58:26 2019 network interface lo: Local host name/address: ::1
Mon Aug 12 09:58:26 2019 network interface eno1: Local host name/address: fe80::f289:46da:cf7b:1d4d
Mon Aug 12 09:58:26 2019 network interface eno1: Local host name/address: logmediapc002
Mon Aug 12 09:58:26 2019 cups-browsed [BrowsePoll /run/cups/cups.sock:0]: IPP-Create-Subscription
Mon Aug 12 09:58:26 2019 cups-browsed [BrowsePoll /run/cups/cups.sock:0]: subscription ID=1746
Mon Aug 12 09:58:26 2019 cups-browsed (/run/cups/cups.sock): cupsEnumDests
Mon Aug 12 09:58:26 2019 find_previous_queue() in THREAD 139740392815168
Mon Aug 12 09:58:26 2019 get-printer-attributes IPP call failed on printer konica_minolta_bizhub_c253_5e_92_c6_ (implicitclass://KONICA_MINOLTA_bizhub_C253_5E_92_C6_/).
Mon Aug 12 09:58:26 2019 ERROR: Unable to create print queue, ignoring printer.
Mon Aug 12 09:58:26 2019 ERROR: Unable to allocate memory.
```

I'm afraid I'm not sure how to find the URL for the ipptool command. Can you hand hold me here? I can see the printer over cups with this url:
http://localhost:631/printers/KONICA_MINOLTA_bizhub_C253_5E_92_C6_ 
I attach a screenshot of it.

![Screenshot from 2019-08-13 08:23:54](https://user-images.githubusercontent.com/17121556/62919717-d1483500-bda3-11e9-88ec-e184e45fc4db.png)

**Comment by deepak0405** [comment]
@logTom 
The cups version is fine, and the fix is not complete yet ( Its not working).  I checked your printer specifications and it only supports IPP 1.1, so I just want to try that whether we are able to create print queue when we set IPP version as 1.1 in the request.

@tillkamppeter 
I tried creating a IPP 1.1 printer on server using the uris
```
         ipp://10.0.1.42/ipp/print?version=1.1
         ipp://localhost:631/resource?version=1.1
```
and I was able to create the print queue. Am I creating the IPP 1.1 printer correctly ? What uri @logTom  should use in ipptool ?

@zdohnal @logTom 
I am not able to create the same environment as yours in my machine. So, Can you please try [this](https://github.com/deepak0405/cups-filters/blob/master/ipp.patch) solution and send me the debug output ?

**Comment by tillkamppeter** [comment]
AFAIK get_printer_attributes works also with IPP 1.1, but you get much less information. As cups-browsed supports also some legacy (IPP 1.1) PostScript/PCL/PDF printers you can also get queues for IPP 1.1 printers, the PPDs are simply not so well fit to the printers (like showing choices which are not available on the actual printer). You can suppress creation of such queues via options in `cups-browsed.conf` and only get queues for "real" driverless IPP printers.
How are you creating printers on the server with the mentioned URIs?

**Comment by DeepakPatankar** [comment]
@tillkamppeter 
I was thinking that since we didn't set the IPP version, cups-browsed is doing request using default version (IPP 2.0), and the printer doesn't understand it, so it was giving error. 

I created the queue in browser, After "Add printer" it asked for the uri, there I entered this uris.

We are not getting HTTP_STATUS_OK for @logTom when we are using httpSeparateURI(), that's why the get_printer_attributes is failing. I am not sure whether both this issues are connected. Things will be more clear once we get the debug output.

**Comment by tillkamppeter** [comment]
> I created the queue in browser, After "Add printer" it asked for the uri, there I entered this uris.
If you create a CUPS queue this way on a server it talks IPP 1.1 to the physical printer, but the shared CUPS queue is understanding IPP 2.x as usual. For a client it does not matter which device URI you have given to a server's CUPS queue, the device URI is only for the communication with the physical printer.

**Comment by DeepakPatankar** [comment]
@tillkamppeter 
How I can add a printer which supports IPP 1.1 ? I saw ippserver but I think there is no option in ippserver to make IPP 1.1 printer.

For @zdohnal  I need a cups queue understanding only IPP 1.1 and for @logTom I need a IPP Network printer.

**Comment by dheeraj135** [comment]
@DeepakPatankar Please try the ippeveprinter utility. Link: https://www.cups.org/doc/man-ippeveprinter.html

**Comment by tillkamppeter** [comment]
@DeepakPatankar (@deepak0405 ? What do you prefer?), note that `ippeveprinter` was only introduced with CUPS 2.3.x. Either use the CUPS packages from Debian Experimental or build from source. Perhaps the [PWG IPP Sample](https://istopwg.github.io/ippsample/) project also has a solution.

**Comment by zdohnal** [comment]
@DeepakPatankar Hi Deepak,
I'm not sure if it is still valid but I was able to get cupsd.conf from our IT. Here it is 
[cupsd.conf.txt](https://github.com/OpenPrinting/cups-filters/files/3525291/cupsd.conf.txt) 

I just substituted IPs, hostnames and paths to certificate/key files with placeholders.

**Comment by deepak0405** [comment]
@tillkamppeter  Please use @deepak0405 .

@zdohnal I am trying to figure out whether your cups version is doing broadcasts using `DNS-SD` or `Cups Legacy Broadcasts`
1. Can you please send the cups-browsed.conf file of the client and the output of `avahi-discover` command when run on client ?
2. Also if you can send the output of `avahi-browse -a` command when run on one of the client whose queue we are not able to generate ?

**Comment by zdohnal** [comment]
@deepak0405 Hi Deepak,
AFAIK the server (cups-1.4.2) is using CUPS legacy broadcast.

Ad 1) Here is cups-browsed.conf 
[cups-browsed.conf.txt](https://github.com/OpenPrinting/cups-filters/files/3533609/cups-browsed.conf.txt) from my machine, where I cannot see any printers when cups-filters 1.24.0 is installed . 
And IMO there will be no avahi communication with server, because they are not in the same network (where DNS-SD supposes to work). Or at least avahi-discover shows nothing :) .

Ad 2) I'm not sure on which machine you want me to run the command - my machine, where I should see queues advertised by server, or on the server?

**Comment by tillkamppeter** [comment]
@deepak0405, the configuration as I understand it sees the server's printers only by legacy CUPS broadcasting and BrowsePoll, most probably it is BrowsePoll which should actually get the printers known to the client.
@zdohnal, Deepak wants you to run the Avahi tools (`avahi-discover`/`avahi-browse`) on the client, to see whether the server's DNS-SD broadcasts make it to the client.

**Comment by deepak0405** [comment]
@zdohnal Thanks now I am able to reproduce the issue. After setting the required ip in `BrowsePoll` I am getting the same error as shown by you.

**Comment by deepak0405** [comment]
@tillkamppeter 
The error I am facing is little bit different 
```
Sun Aug 25 22:17:44 2019 Cannot connect to remote printer with URI ipp://ubuntu:631/printers/test1, ignoring this printer.
Sun Aug 25 22:17:44 2019 get-printer-attributes IPP call failed on printer test1 (ipp://ubuntu:631/printers/test1).

```
I added a BrowsePoll line in the cups-browsed.conf `BrowsePoll <printer-server-ip-address(address of my vm>:631`. The queue was broadcasted but my cups-browsed can't connect to the uri. I tried getting ipp attributes for the uri `ipp://ubuntu:631/printers/test1`, but it says 
```
ipptool: Unable to connect to ubuntu on port 631 - Name or service not known
```
While the request for uri `ipp://<ip-address>:631/printers/test1`, passes without giving any output.

**Log Files**  :
[cups_browsed_output.txt](https://github.com/OpenPrinting/cups-filters/files/3538393/cups_browsed_output.txt)
[printer_attributes_request.txt](https://github.com/OpenPrinting/cups-filters/files/3538394/printer_attributes_request.txt)

**Comment by tillkamppeter** [comment]
In general, if you try to connect to your VM by host name and not by IP you should try to use "ubuntu.local" instead of just "ubuntu".
The ipptool output using the IPP-based address shows that something is wrong with the printer on your server(VM). A regular, shared, and correctly working CUPS queue on a not too old CUPS server gives a useful answer to a get-printer-attributes IPP request. Your VM's ("ubuntu") CUPS queue ("works") gives an empty answer to a remote get-printer-attributes IPP request. This also makes cups-browsed not creating a queue. cups-browsed does a get-printer-attributes IPP request and uses that information to create a local print queue which correctly prints on the remote printer, assuming cups-browsed discovers the printer by IPP.
This can be the problem of @zdohnal's CUPS server. Assuming that that server and your VM use the same CUPS version and the same CUPS configuration now @zdohnal's server also gives empty answers to get-printer-attributes IPP requests and so a client cannot use the server's printer solely based on DNS-SD and IPP.
One could try to configure CUPS 1.4.2 to correctly support DNS-SD/IPP-based clients but if it does not work out (not getting useful answers to get-printer-attributes IPP requests) we cannot expect upstream or Red Hat to fix it.
The next option would be to check whether cups-browsed works correctly with CUPS legacy browsing. If it discovers a printer only by either BrowsePoll or by CUPS legacy broadcasts from the server, does it correctly set up a queue based on a PPD file downloaded from the server? And if cups-browsed discovers a printer by BrowsePoll or by CUPS legacy broadcasts AND also by DNS-SD but the get-printer-attributes IPP request fails as observed here, does cups-browsed then create a local queue using PPD download?
For the further development of cups-browsed we need to note the following:
- A completely PPD-less approach only works for printers discovered by DNS-SD and answering get-printer-attributes IPP requests correctly, In this case we should not download the server's PPD, base ourselves on the get-printer-attributes IPP requests and do clustering. In this case we fully conform with PPD deprecation on the discovery end.
- We cannot remove the functionality for downloading the PPD from the server in cups-browsed, in case of a printer discovered by legacy (BrowsePoll or server broadcast) we need to download the PPD, also if discovered by both legacy and DNS-SD but get-printer-attributes IPP request failing.

This is how cups-browsed should work for now.

**Comment by logTom** [comment]
@deepak0405
Sorry for the late response.
I downloaded the cups-filters 1.25.4 release and applied your patch.

Here is the output:
```
Mon Sep  2 09:05:58 2019 Reading config: BrowseRemoteProtocols dnssd cups
Mon Sep  2 09:05:58 2019 No "Browse..." line at all, accept all servers ("BrowseOrder Deny,Allow").
Mon Sep  2 09:05:58 2019 main() in THREAD 140343734065728
Mon Sep  2 09:05:58 2019 cups-browsed: Creating http connection to local CUPS daemon: /var/run/cups/cups.sock:631
Mon Sep  2 09:05:58 2019 update_netifs() in THREAD 140343734065728
Mon Sep  2 09:05:58 2019 network interface lo: Local host name/address: 127.0.0.1
Mon Sep  2 09:05:58 2019 network interface lo: Local host name/address: localhost
Mon Sep  2 09:05:58 2019 network interface enp0s3: Local host name/address: 192.168.1.55
Mon Sep  2 09:05:59 2019 network interface enp0s3: Local host name/address: loguservm.local
Mon Sep  2 09:05:59 2019 Network interface enp0s3 at 192.168.1.55 for legacy CUPS browsing/broadcasting
Mon Sep  2 09:05:59 2019 network interface lo: Local host name/address: ::1
Mon Sep  2 09:05:59 2019 network interface enp0s3: Local host name/address: fe80::1a8e:b96e:77e6:f6af
Mon Sep  2 09:05:59 2019 network interface enp0s3: Local host name/address: loguservm
Mon Sep  2 09:05:59 2019 cups-browsed [BrowsePoll /var/run/cups/cups.sock:0]: IPP-Create-Subscription
Mon Sep  2 09:05:59 2019 cups-browsed [BrowsePoll /var/run/cups/cups.sock:0]: subscription ID=26
Mon Sep  2 09:05:59 2019 cups-browsed (/var/run/cups/cups.sock): cupsEnumDests
Mon Sep  2 09:05:59 2019 find_previous_queue() in THREAD 140343734065728
Mon Sep  2 09:05:59 2019 ERROR: Invalid URI: implicitclass://KONICA_MINOLTA_bizhub_C253_5E_92_C6_/Mon Sep  2 09:05:59 2019 get-printer-attributes IPP call failed on printer konica_minolta_bizhub_c253_5e_92_c6_ (implicitclass://KONICA_MINOLTA_bizhub_C253_5E_92_C6_/).
Mon Sep  2 09:05:59 2019 ERROR: Unable to create print queue, ignoring printer.
```

**Comment by logTom** [comment]
I tried out the latest cups-filters 1.25.5 release. The output is different now, but still doesn't work:

```
sudo /usr/sbin/cups-browsed -v
Passwort: 
Mon Sep  9 10:27:26 2019 Reading command line option -v, turning on debug mode (Log on standard error).
Mon Sep  9 10:27:26 2019 cups-browsed of cups-filters version 1.25.5 starting.
Mon Sep  9 10:27:26 2019 Reading config: BrowseRemoteProtocols dnssd cups
Mon Sep  9 10:27:26 2019 No "Browse..." line at all, accept all servers ("BrowseOrder Deny,Allow").
Mon Sep  9 10:27:26 2019 main() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 cups-browsed: Creating http connection to local CUPS daemon: /var/run/cups/cups.sock:631
Mon Sep  9 10:27:26 2019 update_netifs() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 network interface lo: Local host name/address: 127.0.0.1
Mon Sep  9 10:27:26 2019 network interface lo: Local host name/address: localhost
Mon Sep  9 10:27:26 2019 network interface enp0s3: Local host name/address: 192.168.1.56
Mon Sep  9 10:27:26 2019 network interface enp0s3: Local host name/address: loguservm.local
Mon Sep  9 10:27:26 2019 Network interface enp0s3 at 192.168.1.56 for legacy CUPS browsing/broadcasting
Mon Sep  9 10:27:26 2019 network interface lo: Local host name/address: ::1
Mon Sep  9 10:27:26 2019 network interface enp0s3: Local host name/address: fe80::1a8e:b96e:77e6:f6af
Mon Sep  9 10:27:26 2019 network interface enp0s3: Local host name/address: loguservm
Mon Sep  9 10:27:26 2019 cups-browsed [BrowsePoll /var/run/cups/cups.sock:0]: IPP-Create-Subscription
Mon Sep  9 10:27:26 2019 cups-browsed [BrowsePoll /var/run/cups/cups.sock:0]: subscription ID=39
Mon Sep  9 10:27:26 2019 cups-browsed (/var/run/cups/cups.sock): cupsEnumDests
Mon Sep  9 10:27:26 2019 find_previous_queue() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 === Remote printer overview ===
Mon Sep  9 10:27:26 2019 ===============================
Mon Sep  9 10:27:26 2019 === Remote printer overview ===
Mon Sep  9 10:27:26 2019 Printer  (, ): Local queue konica_minolta_bizhub_c253_5e_92_c6_, Remote CUPS Printer, Slave of None (To be created/updated)
Mon Sep  9 10:27:26 2019 ===============================
Mon Sep  9 10:27:26 2019 Found CUPS queue konica_minolta_bizhub_c253_5e_92_c6_ (URI: ) from previous session.
Mon Sep  9 10:27:26 2019 Using signal handler SIGACTION
Mon Sep  9 10:27:26 2019 Avahi server connection got available, setting up service browsers.
Mon Sep  9 10:27:26 2019 checking queues in 300s
Mon Sep  9 10:27:26 2019 browse_callback() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 Avahi Browser: NEW: service 'KONICA MINOLTA bizhub C253(5E:92:C6)' of type '_ipp._tcp' in domain 'local' on interface 'enp0s3' (IPv6)
Mon Sep  9 10:27:26 2019 browse_callback() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 Avahi Browser: NEW: service 'KONICA MINOLTA bizhub C253(5E:92:C6)' of type '_ipp._tcp' in domain 'local' on interface 'enp0s3' (IPv4)
Mon Sep  9 10:27:26 2019 browse_callback() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 Unable to find interface name for interface -1: No such device or address
Mon Sep  9 10:27:26 2019 Avahi Browser: CACHE_EXHAUSTED
Mon Sep  9 10:27:26 2019 browse_callback() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 Unable to find interface name for interface -1: No such device or address
Mon Sep  9 10:27:26 2019 Avahi Browser: ALL_FOR_NOW
Mon Sep  9 10:27:26 2019 browse_callback() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 Unable to find interface name for interface -1: No such device or address
Mon Sep  9 10:27:26 2019 Avahi Browser: CACHE_EXHAUSTED
Mon Sep  9 10:27:26 2019 browse_callback() in THREAD 140541639387712
Mon Sep  9 10:27:26 2019 Unable to find interface name for interface -1: No such device or address
Mon Sep  9 10:27:26 2019 Avahi Browser: ALL_FOR_NOW
Mon Sep  9 10:27:27 2019 resolve_callback() in THREAD 140541639387712
Mon Sep  9 10:27:27 2019 update_netifs() in THREAD 140541639387712
Mon Sep  9 10:27:27 2019 network interface lo: Local host name/address: 127.0.0.1
Mon Sep  9 10:27:27 2019 network interface lo: Local host name/address: localhost
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: 192.168.1.56
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: loguservm.local
Mon Sep  9 10:27:27 2019 Network interface enp0s3 at 192.168.1.56 for legacy CUPS browsing/broadcasting
Mon Sep  9 10:27:27 2019 network interface lo: Local host name/address: ::1
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: fe80::1a8e:b96e:77e6:f6af
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: loguservm
Mon Sep  9 10:27:27 2019 Avahi Resolver: Service 'KONICA MINOLTA bizhub C253(5E:92:C6)' of type '_ipp._tcp' in domain 'local' with host name 'KMBT5E92C6.local' and port 631 on interface 'enp0s3' (IPv6).
Mon Sep  9 10:27:27 2019 Avahi Resolver: Service 'KONICA MINOLTA bizhub C253(5E:92:C6)' of type '_ipp._tcp' in domain 'local' skipped, not a local service.
Mon Sep  9 10:27:27 2019 checking queues in 299s
Mon Sep  9 10:27:27 2019 resolve_callback() in THREAD 140541639387712
Mon Sep  9 10:27:27 2019 update_netifs() in THREAD 140541639387712
Mon Sep  9 10:27:27 2019 network interface lo: Local host name/address: 127.0.0.1
Mon Sep  9 10:27:27 2019 network interface lo: Local host name/address: localhost
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: 192.168.1.56
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: loguservm.local
Mon Sep  9 10:27:27 2019 Network interface enp0s3 at 192.168.1.56 for legacy CUPS browsing/broadcasting
Mon Sep  9 10:27:27 2019 network interface lo: Local host name/address: ::1
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: fe80::1a8e:b96e:77e6:f6af
Mon Sep  9 10:27:27 2019 network interface enp0s3: Local host name/address: loguservm
Mon Sep  9 10:27:27 2019 Avahi Resolver: Service 'KONICA MINOLTA bizhub C253(5E:92:C6)' of type '_ipp._tcp' in domain 'local' with host name 'KMBT5E92C6.local' and port 631 on interface 'enp0s3' (IPv4).
Mon Sep  9 10:27:27 2019 Avahi Resolver: Service 'KONICA MINOLTA bizhub C253(5E:92:C6)' of type '_ipp._tcp' in domain 'local' skipped, not a local service.
Mon Sep  9 10:27:27 2019 checking queues in 299s
Mon Sep  9 10:32:26 2019 Processing printer list ...
Mon Sep  9 10:32:26 2019 === Remote printer overview ===
Mon Sep  9 10:32:26 2019 Printer  (, ): Local queue konica_minolta_bizhub_c253_5e_92_c6_, Remote CUPS Printer, Slave of None (Unconfirmed)
Mon Sep  9 10:32:26 2019 ===============================
Mon Sep  9 10:32:26 2019 No remote printer named konica_minolta_bizhub_c253_5e_92_c6_ available, removing entry from previous session.
Mon Sep  9 10:32:26 2019 Printer konica_minolta_bizhub_c253_5e_92_c6_ (Host: , Port: 631, URI: ) disappeared and no slave available (or it is a slave of another printer), removing entry.
Mon Sep  9 10:32:26 2019 Removing entry konica_minolta_bizhub_c253_5e_92_c6_ () and its CUPS queue.
Mon Sep  9 10:32:26 2019 Recording printer options for konica_minolta_bizhub_c253_5e_92_c6_ to /var/cache/cups/cups-browsed-options-konica_minolta_bizhub_c253_5e_92_c6_
Mon Sep  9 10:32:26 2019 Printer konica_minolta_bizhub_c253_5e_92_c6_ got modified externally, discovered by the NickName of its PPD file having changed from "(no PPD)" to "ONICA MINOLTA C25, driverless, cups-filters 1.25.1".
Mon Sep  9 10:32:26 2019 === Remote printer overview ===
Mon Sep  9 10:32:26 2019 ===============================
Mon Sep  9 10:32:26 2019 listening
```

**Comment by logTom** [comment]
Is there anything I can do to get "driverless printing" working again and thus get this issue resolved?

**Comment by paulmenzel** [comment]
@logTom, as the output is different, please open a separate issue (and also test the release 1.25.6).

@zdohnal, can you still reproduce the issue?

**Comment by zdohnal** [comment]
I'll build the newest version and let you know shortly.

**Comment by zdohnal** [comment]
Unfortunately, the errors are still there and cups-browsed does not create any local queues.

The current debug log 
[cups_browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3680618/cups_browsed_log.txt)
but the errors are the same or similar:

Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Creating/Updating CUPS queue BRQ-TPBC-1D124-North
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Creating permanent CUPS queue BRQ-TPBC-1D124-North.
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Loading saved printer options for BRQ-TPBC-1D124-North from /var/cache/cups/cups-browsed-options-BRQ-TPBC-1D124-North
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Failed reading file /var/cache/cups/cups-browsed-options-BRQ-TPBC-1D124-North, probably no options recorded yet
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Print queue BRQ-TPBC-1D124-North is for remote CUPS queue(s) and we get notifications from CUPS, using implicit class device URI implicitclass://BRQ-TPBC-1D124-North/
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Request for IPP attributes (get-printer-attributes) for printer with URI ipp://server_hostname:631/printers/BRQ-TPBC-1D124-North failed: Success
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 get-printer-attributes IPP call failed on printer BRQ-TPBC-1D124-North (ipp://server_hostname:631/printers/BRQ-TPBC-1D124-North).
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Removing entry BRQ-TPBC-1D124-North (ipp://server_hostname:631/printers/BRQ-TPBC-1D124-North) and its CUPS queue.
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Recording printer options for BRQ-TPBC-1D124-North to /var/cache/cups/cups-browsed-options-BRQ-TPBC-1D124-North
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Unable to get PPD file for BRQ-TPBC-1D124-North: The printer or class does not exist.
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 lpstat: No destinations added.
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Unable to load PPD from local queue BRQ-TPBC-1D124-North, queue seems to be raw.
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 Queue has still jobs or CUPS error!
Oct 02 06:10:44 host-10-0-137-198 cups-browsed[16205]: Wed Oct  2 06:10:44 2019 ERROR: Failed disabling printer 'BRQ-TPBC-1D124-North': The printer or class does not exist.

**Comment by deepak0405** [comment]
Hey @paulmenzel 
We have not shipped the solution for this issue yet. I am occupied a lot by my day job, and whatever time I am getting is going in solving the other bugs.

Hey @logTom,
Do you want to contribute to solving this issue? I know how we can fix it and will love to help you. I think the work will take something around 3-5 days.

**Comment by logTom** [comment]
@deepak0405
Sure, you can contact me via email.

**Comment by deepak0405** [comment]
Hey @logTom 
Thank You

**Comment by logTom** [comment]
@deepak0405 I'm glad if I can somehow help this awesome project. 
My work mates using windows are always blown away by the fact that I'm able to print stuff without having to manually install any printer drivers.

**Comment by opoplawski** [comment]
@logTom - I seem to be seeing similar errors (Unable to find interface name for interface -1: No such device or address) - are you planning on opening a separate issue as @paulmenzel suggested?

**Comment by logTom** [comment]
@opoplawski, @paulmenzel: Here we go #163 - I will also test 1.25.6.

**Comment by deepak0405** [comment]
@zdohnal 
I have implemented the fallback logic for the IPP1.1 request.Can you please check whether [this](https://github.com/deepak0405/cups-filters/tree/do-fallback-for-IPP1.1) branch solves your issue ?

**Comment by zdohnal** [comment]
Hi @deepak0405 ,
thank you for your hard work! I'll check it today.

**Comment by zdohnal** [comment]
@deepak0405 
Unfortunately, your current code segfaults due infinite recursion. If I forbid it after the first IPP1.1 attempt (add if/else whether to recurse deeper or return NULL), it does not crash, but it does not add print queues neither.
Logs: 
[log.txt](https://github.com/OpenPrinting/cups-filters/files/3707646/log.txt)

I tried to put back some of older cups-browsed code, where cups-browsed gets PPD from server and create print queue, but I got stuck with different implementation of implicitclass backend (the old one connected to server and created job by IPP-CREATE-JOB op, but now the backend uses other binary to put the data through).

**Comment by tillkamppeter** [comment]
I have fixed the infinite recursion now. Please test the current GIT repo of cups-filters.

**Comment by zdohnal** [comment]
Unfortunately, no change neither. Log 
[browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3708447/browsed_log.txt)

**Comment by tillkamppeter** [comment]
Does this mean that you still get the infinite recursion?

**Comment by zdohnal** [comment]
I'm sorry, I'll be precise - no infinite recursion this time, but print queues are not created too.

**Comment by zdohnal** [comment]
And I discovered two memory leaks - I created pull request #165 for fixing it.

**Comment by tillkamppeter** [comment]
Already merged, thank you very much.

**Comment by tillkamppeter** [comment]
@zdohnal, I have done some more fixes, can you test with the current state of cups-filters? Thanks.

**Comment by zdohnal** [comment]
@tillkamppeter Np, I'll check it right away, thank you!

**Comment by zdohnal** [comment]
I'm sorry, it does not work still :( here is the log 
[cups-browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3712938/cups-browsed_log.txt)

**Comment by deepak0405** [comment]
@zdohnal @tillkamppeter 
I was able to reproduce the error because of some different reason. I am trying to reproduce the exact error, Can you please help?

I have the latest cups-filters on my machine, and on a VM, I have installed cups-1.4.2. I created a cups queue on VM using
```
lpadmin -p test2 -m foomatic-db-compressed-ppds:0/ppd/foomatic-ppd/Xerox-WorkCentre_Pro_133-Postscript.ppd -o printer-is-shared=true 
```
If I don't use BrowsePoll line, the queue is not made, and after adding the BrowsePoll line I am getting error.
```
Could not connect to the server ubuntu:631 for test2!
```
The cups-browsed is not able to connect to the printer for getting the attributes. In case of @zdohnal , the cups-browsed is able to connect to the printer. Am I missing some configuration here ? I also tried running a previous version of cups-filters(1.17.9), but was getting the same error.

**Comment by tillkamppeter** [comment]
@deepak0405, you have really the very latest cups-filters (1.25.11 or top of GIT repo)?

**Comment by deepak0405** [comment]
@tillkamppeter, Yes till I am using the latest 1.25.11 version

**Comment by tillkamppeter** [comment]
@deepak0405, does CUPS 1.4.2 support IPP 2.x? If not, does the get-printer-attributes IPP-1.1 fallback work correctly?
I CUPS 1.4.2 DNS-SD-broadcasting printers? Or does it only advertise via legacy CUPS broadcasting?

**Comment by tillkamppeter** [comment]
If a printer gets discovered by legacy CUPS broadcasting (or by BrowsePoll) does cups-browsed have enough info to IPP-poll this printer and does it the IPP poll correctly then?

**Comment by deepak0405** [comment]
@tillkamppeter By IPP-poll here you mean getting the IPP attributes of the printer?
The cups-browsed is not able to connect to the printer and thus not able to get any attributes.

In case of cups-filters 1.17,9, the cups-browsed tries to download the ppd file from the server, but in that case also, I was not able to connect to the printer.

**Comment by tillkamppeter** [comment]
@deepak0405, yes, I mean the get-printer-attributes IPP request.
Is it not able to create an HTTP connection to the CUPS daemon on the VM? Or does it create this connection but somehow the IPP request fails?
Van you access the VM's CUPS daemon via the command line on your host machine, using the `-h` option:
```
lpstat -h ubuntu:631 -v
lpoptions -h ubuntu:631 -p test2 -l
```

**Comment by tillkamppeter** [comment]
Is the CUPS on the VM configured to listen to remote hosts?
On the VM run
```
cupsctl
cupsctl --share-printers
```
On the client run
```
nmap ubuntu
```
Is port 631 of the VM shown as open?

**Comment by deepak0405** [comment]
It is not able to make a http connection. The print queues are broadcasted. 
Output of commands:
lpstat -h ubuntu:631 -v
``` lpstat: Resource temporarily unavailable
```

lpoptions -h ubuntu:631 -p test2 -l
```
lpoptions: Unable to get PPD file for test2: No such file or directory
```

When I run cups 2.10 on the server, the queue is generated correctly and the IPP request also works fine. The output of `cupsctl` shows that the printer is shared.

nmap ubuntu 
```
Starting Nmap 7.70 ( https://nmap.org ) at 2019-10-13 02:22 IST
Failed to resolve "ubuntu".
WARNING: No targets were specified, so 0 hosts scanned.
Nmap done: 0 IP addresses (0 hosts up) scanned in 0.04 seconds
```

Yes the port is open.

**Comment by tillkamppeter** [comment]
It seems that CUPS 1.4.2 on your VM is generally not accessible for a remote machine. Have you checked cupsd.conf on your VM whether all settings are correct?

**Comment by tillkamppeter** [comment]
@deepak0405, also note that I am assuming that your VM's host name is "ubuntu", following your examples. If it is not the case or needs something extra, like "ubuntu.local", please substitute the "ubuntu" in my examples for whatever is correct as a host name or IP for your VM.

**Comment by deepak0405** [comment]
On my vm I am running cups in a different directory, i.e. I used `./configure --prefix=/<some-path>` to build the cups. Can that be the reason ?

**Comment by deepak0405** [comment]
With ip-address, one of the command works
```
deepak@deepak-Lenovo-G50-80:/etc/cups$ lpstat -h 192.168.2xx.1xx:631 -v
device for test2: ///dev/null
device for works: ipp://localhost:631/ipp/resource?version=1.1
```
but this one doesn't
```
deepak@deepak-Lenovo-G50-80:/etc/cups$ lpoptions -h 192.168.2xx.1xx:631 -p test2 -l
lpoptions: Unable to get PPD file for test2: Name or service not known

```

**Comment by tillkamppeter** [comment]
Installing into an alternative directory on the VM is no problem as long as you do not have the original CUPS running in parallel.
The `lpstat` command with IP shows that CUPS is principally accessible on the server, but perhaps you did not use the correct host name when accessing with a host-name-based URI.
That the `lpoptions` command with IP does not work is strange. Perhaps you should have a look into the source code of `lpstat` and `lpoptions` (of the CUPS version running on your client) and also have a look into the CUPS error_log on your VM.

**Comment by tillkamppeter** [comment]
The error message of `lpoptions` is caused by this code piece in the list_options() function:
```
  if ((http = cupsConnectDest(dest, CUPS_DEST_FLAGS_NONE, 30000, NULL, resource,
 sizeof(resource), NULL, NULL)) == NULL)
  {
    _cupsLangPrintf(stderr, _("lpoptions: Unable to get PPD file for %s: %s"),
                    dest->name, cupsLastErrorString());
    return;
  }

  if ((filename = cupsGetPPD2(http, dest->name)) == NULL)
  {
    httpClose(http);

    _cupsLangPrintf(stderr, _("lpoptions: Unable to get PPD file for %s: %s"),
                    dest->name, cupsLastErrorString());
    return;
  }
```
Unfortunately, both steps have the same error message if they fail, so only modifying this piece of code (on the client) can tell you where it exactly fails. But for me it looks like that it is the second one, because the `lpstat` works and `lpstat` probably has the same first but a different second step. My guess here is that `lpstat` speaks IPP after creating the HTTP connection while `lpoptions` downloads the PPD, using HTTP.

**Comment by deepak0405** [comment]
I am able to do `get-printer-attributes` request using the IP address of the vm, so I just changed the name `ubuntu` with the ip in my cups-browsed.c and finally I was able to reproduce the error.

The `backend/ipp.c` handles the IPP request nicely, it also does fallback request
```
       else if ((ipp_status == IPP_STATUS_ERROR_BAD_REQUEST ||
              ipp_status == IPP_STATUS_ERROR_VERSION_NOT_SUPPORTED) && version > 10)
       // IPP2.0 not working trying IPP 1.1
```
In case of @logTom, we are getting the error IPP_STATUS_ERROR_VERSION_NOT_SUPPORTED and thus we are doing a fallback IPP1.1 request, while in case of @zdohnal, the ipp_status is null( but it should not be) and no data is returned by the printer. But if we do a request without media-col-database, the printer returns the required info. @zdohnal  Please try [this](https://github.com/OpenPrinting/cups-filters/compare/master...deepak0405:printing-in-IPP1.1?expand=1) dummy branch, for my case I was able to create a cups queue with sufficient printer info.

@zdohnal  I am not quite sure whether the printing will work after it, as it is not working for @logTom, but I think this PR should be able to generate ppd files and job should also be sent.

@tillkamppeter If this branch works, we can follow this strategy :
If the request return ipp_status one of the two shown above, we will do a IPP1.1 request.
If the get-attributes-call is not having sufficient attributes( like which happens in @zdohnal's case), we do a fallback request without "media-col-database".

@tillkamppeter @zdohnal I don't think there is need to revert back to previous code of downloading a ppd file for legacy printer, as in case of @zdohnal  and @logTom, we get sufficient output for get-printer-attributes request.

**Comment by tillkamppeter** [comment]
@deepak0405, OK, let us go this way.

**Comment by zdohnal** [comment]
@deepak0405 The print queues are now visible. Thank you, Deepak!

Although it takes about 20-30s to 1st print queue to show up now.

Regarding printing, the job ends with:

implicitclass://BRQ-TPBC-1D124-North/: Unable to create job - No such file or directory. Retrying.

**Comment by deepak0405** [comment]
Sorry for the late response, I was not well the last week.

I tried reproducing the error, but the printing is working for my setup. Can you please attach the error logs ?

**Comment by zdohnal** [comment]
@deepak0405  No problem, Deepak. I hope you feel better.

Here are logs for the job 
[cupsd_job_log.txt](https://github.com/OpenPrinting/cups-filters/files/3750959/cupsd_job_log.txt) . Actually previous error was caused by old implicitclass backend. Now logs are saying the printing is okay, but only paper which got out of printer is on the photo

![IMG_20191021_115004](https://user-images.githubusercontent.com/17822565/67208804-7ae81a00-f416-11e9-8bfd-2666b96730df.jpg)

**Comment by paulmenzel** [comment]
This issues has become too long. @zdohnal, can you please create a new one, and reference it here?

**Comment by deepak0405** [comment]
@tillkamppeter This PR #173 contains the complete implementation of the fallback logic. I hope after this PR, we can close this issue, since queues will be generated for all the cases. But the issue for printing will still remain open, and I am currently working on it.

@zdohnal @logTom  Can you please verify this PR ?

**Comment by logTom** [comment]
@deepak0405 Yeah, I tested it. But, in my case the queues seem to be fixed already before the pr. Printing still doesn't work. I guess you only need the logs if it still doesn't work for zdohnal.

**Comment by tillkamppeter** [comment]
@logTom, as printing does not work for you and this PR is about making printing work, we need also your logs. Can you please attach them? Both the CUPS error_log and the cups-browsed_log?

**Comment by zdohnal** [comment]
@deepak0405 Ok, I'll test it on clean cups-filters without your dummy branch.

I hope I'll manage it today, but tomorrow is more possible.

**Comment by zdohnal** [comment]
@deepak0405 Ok, I can see the first print queue now after 1m 20s - cups-browsed 1.22.5 was able to found and create the first queue in 3s.

The printing still does not work, as you expected.

**Comment by deepak0405** [comment]
@zdohnal  1m is too much time, let me see why its taking this much. For your case we will be doing the request thrice
1. First time a IPP 2.0 request.
2. Next time a IPP1.1 request.
3. If both doesn't give good result we do a request without media-col-database.

I was expecting some delay, but not this much. Can you please attach the cups-error-log.

**Comment by tillkamppeter** [comment]
@zdohnal, please also attach the cups-browsed_log.

**Comment by zdohnal** [comment]
@tillkamppeter  @deepak0405 Here you are 
[cups-browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3832308/cups-browsed_log.txt)

**Comment by zdohnal** [comment]
Note: I stopped the service after the first queue came up, so the log should be more relevant than log of whole process.

**Comment by tillkamppeter** [comment]
I have fixed several bugs in cups-filters now. Could you try the newest GIT snapshot? Note that there is no workaround for a Poppler segfault on the server (as you mention in issue #163).

**Comment by tillkamppeter** [comment]
To solve issue #181 I have done several fixes and improvements on the implicitclass backend. Please re-test to see whether this also solves this bug.

**Comment by tillkamppeter** [comment]
@logTom, @zdohnal, I have now restructured the get_printer_attributes() function of cups-browsed and moved it into libcupsfilters, to share it with the driverless utility (commit 20c4773).
This could perhaps also fix this issue, please test (if the cause of this issue is not a segfaulting Poppler on the server).

**Comment by zdohnal** [comment]
@tillkamppeter I'm able to print with cups-filters-1.26.0 after fixing poppler. When I set:

```
BrowsePoll <el6_print_server>
UpdateCUPSQueuesMaxPerCall 20
PauseBetweenCUPSQueueUpdates 5
```

the first queue of 124 queues appeared after 55s - here is the cups-browsed_log 
[cups-browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3967825/cups-browsed_log.txt) .

It is better than previous interval though, but I'm not sure if it is expected.

**Comment by tillkamppeter** [comment]
This means that this issue is fixed. The slowness of queue creation needs to get investigated, a near-future feature request.

**Comment by tillkamppeter** [comment]
@zdohnal, what makes it take 55 seconds before the first queue is created is the fact that all your 124 printers are BrowsePolled from one server, so they are discovered in one single event response (timeout event to trigger a BrowsePoll on a given server) and in this single event response all entries for the local queues are generated in cups-browsed's internal data structure. For queues to be created, the current event response must terminate, so that another one (the queue creation, triggered by the earliest timeout in the printer entries) can start. So what has to be improved is that also BrowsePoll event responses can be split in smaller portions, like we already do with the CUPS queue update event response.

**Comment by zdohnal** [comment]
@tillkamppeter Ok, I'll open the new issue for it to track it.

Thank you for explanation and fixing the issue!
